crates/ruff_linter/src/rules/fastapi/rules/fastapi_unused_path_parameter.rs

```diff
@@ -0,0 +1,198 @@
+use crate::fix::edits::add_parameter;
+use crate::rules::fastapi::rules::is_fastapi_route;
+use crate::rules::fastapi::rules::is_fastapi_route_decorator;
+use regex::Regex;
+use ruff_python_stdlib::identifiers::is_identifier;
+
+/// ## What it does
+/// Identifies FastAPI routes that declare path parameters in the route path but not in the function signature.
+///
+/// ## Why is this bad?
+/// Path parameters are used to extract values from the URL path.
+/// If a path parameter is declared in the route path but not in the function signature, it will not be accessible in the function body, which is likely a mistake.
+///
+/// ## Known problems
+/// If the path parameter is not a valid Python identifier, FastAPI will normalize it to a valid identifier.
+/// This lint simply ignores path parameters that are not valid identifiers, as that normalization behavior is undocumented.
+///
+/// ## Example
+///
+/// ```python
+/// from fastapi import FastAPI
+///
+/// app = FastAPI()
+///
+///
+/// @app.get("/things/{thing_id}")
+/// async def read_thing(query: str):
+///     ...
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// from fastapi import FastAPI
+///
+/// app = FastAPI()
+///
+///
+/// @app.get("/things/{thing_id}")
+/// async def read_thing(thing_id: int, query: str):
+///     ...
+/// ```
+
+#[violation]
+pub struct FastApiUnusedPathParameter {
+    arg_name: String,
+    function_name: String,
+    arg_name_already_used: bool,
+}
+
+impl Violation for FastApiUnusedPathParameter {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self {
+            arg_name,
+            function_name,
+            arg_name_already_used: arg_name_is_pos_only_arg,
+        } = self;
+        if *arg_name_is_pos_only_arg {
+            format!(
+                "Path parameter `{arg_name}` in route path, but appears as a positional-only argument in function `{function_name}`. Consider making it a regular argument."
+            )
+        } else {
+            format!(
+                "Path parameter `{arg_name}` in route path but not in function `{function_name}`."
+            )
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let Self { arg_name, .. } = self;
+        if self.arg_name_already_used {
+            None
+        } else {
+            Some(format!("Add `{arg_name}` to function signature"))
+        }
+    }
+}
+
+/// Returns a vector of path parameters and their ranges in the route path.
+/// The string is just the name of the path parameter.
+/// The range includes the curly braces.    
+fn extract_path_params_from_route(input: &str) -> Vec<(String, Range<usize>)> {
+    // We ignore text after a colon, since those are path convertors
+    // See also: https://fastapi.tiangolo.com/tutorial/path-params/?h=path#path-convertor
+    let re = Regex::new(r"\{([^:}]+).*?\}").unwrap();
```

Comment:
Regular expressions are fairly expensive to construct. That's why we should try to only construct them once by e.g. using Lazy

https://github.com/astral-sh/ruff/blob/fe7d965334e6299d23ffbaee9c296edd58e0f1c5/crates/ruff_python_formatter/tests/normalizer.rs#L63-L77

But I think we can do even better here and avoid using a regex all together. It seems all we need is to find a `:` and then take everything up to `}`. This can be done by first doing two `split_once` calls (first by `:`, then by `}`)